core/thread/src/TReentrantRWLock.cxx

```diff
@@ -221,100 +221,151 @@ void TReentrantRWLock<MutexT, RecurseCountsT>::WriteUnLock(TVirtualRWMutex::Hint
+      Error("TReentrantRWLock::Rewind", "Inconsistent read lock count!");
+      return nullptr;
+   }
 
-      // Now set the lock (and potential read locks) for immediate release.
-      fReaders -= readerCount;
-      fRecurseCounts.fWriteRecurse = 1;
+   if (pStateDelta->fDeltaWriteRecurse < 0) {
+      Error("TReentrantRWLock::Rewind", "Inconsistent write lock count!");
+      return nullptr;
+   }
 
-      *(pState->fReadersCountLoc) = 0;
+   auto hint = reinterpret_cast<TVirtualRWMutex::Hint_t *>(typedState.fReadersCountLoc);
+   if (pStateDelta->fDeltaWriteRecurse != 0) {
+      // Notify through WriteUnLock()
+      *typedState.fReadersCountLoc = typedState.fReadersCount;
+      fRecurseCounts.fWriteRecurse = typedState.fWriteRecurse - 1;
 
       // Release this thread's write lock
-      WriteUnLock(reinterpret_cast<TVirtualRWMutex::Hint_t *>(pState->fReadersCountLoc));
-   } else if (readerCount) {
-      // Now set the lock for release.
-      fReaders -= (readerCount-1);
-      *(pState->fReadersCountLoc) = 1;
+      WriteUnLock(hint);
+   } else if (pStateDelta->fReadersCountLoc != 0) {
+      // Notify through ReadUnLock()
+      *typedState.fReadersCountLoc = typedState.fReadersCount - 1;
+      fRecurseCounts.fWriteRecurse = typedState.fWriteRecurse;
 
       // Release this thread's reader lock(s)
-      ReadUnLock(reinterpret_cast<TVirtualRWMutex::Hint_t *>(pState->fReadersCountLoc));
+      ReadUnLock(hint);
    }
-
-   // Do something.
-   //::Fatal("Reset()", "Not implemented, contact pcanal@fnal.gov");
-   return std::move(pState);
+   // else earlierState and *this are identical!
+   return pStateDelta;
 }
 
+//////////////////////////////////////////////////////////////////////////
+/// Re-apply a delta.
+
 template <typename MutexT, typename RecurseCountsT>
-void TReentrantRWLock<MutexT, RecurseCountsT>::Restore(std::unique_ptr<TVirtualMutex::State> &&state)
-{
-   TReentrantRWLockState<MutexT, RecurseCountsT> *pState = dynamic_cast<TReentrantRWLockState<MutexT, RecurseCountsT> *>(state.get());
-   if (!pState) {
-      if (state) {
-         SysError("Restore", "LOGIC ERROR - invalid state object!");
-         return;
-      }
-      // No state, do nothing.
+void TReentrantRWLock<MutexT, RecurseCountsT>::Apply(std::unique_ptr<StateDelta> &&state) {
+   if (!state) {
+      Error("TReentrantRWLock::Apply", "Cannot apply empty delta!");
       return;
    }
 
-   // At a restore point, this thread should not be holding any part
-   // of the lock (if it does the following code will forget about it)
-   assert( *(pState->fReadersCountLoc) == 0);
-   // assert( auto local = fRecurseCounts.GetLocal() && fRecurseCounts.IsNotCurrentWriter(local) )
+   using StateDelta_t = TReentrantRWLockStateDelta<MutexT, RecurseCountsT>;
+   const StateDelta_t* typedDelta = static_cast<const StateDelta_t*>(state.get());
 
-   const auto readerCount = pState->fReadersCount;
+   if (typedDelta->fDeltaWriteRecurse < 0) {
```

Comment:
The locking code is 'performance critical'.  Do we really need the if statement all the time?  Can they be #if DEBUG-ed out?

Reply:
My concern is user code like
```
m.WriteLock();
m.WriteLock();
m.WriteLock();
auto s = m.GetStateBefore(); // s has write recurse count == 2
m.Unlock();
m.Unlock();
m.Rewind(s); // going from write recurse count == 1 to 2
```
which isn't handled at the moment.